chore(babel-plugin-remove-graphql-queries): Convert index.js

zachtylr21 opened the pull request:
## Description

Convert `packages/babel-plugin-remove-graphql-queries/src/index.js` to TypeScript

## Related Issues

Related to #21995

pieh commented on packages/babel-plugin-remove-graphql-queries/src/index.ts:
```diff
@@ -139,23 +206,23 @@ function removeImport(tag) {
 
   const text = quasis[0].value.raw
-  const hash = murmurhash(text, `abc`)
+  const hash = murmurhash(text, 0)
```
That's interesting. I get why we need `number` here (`murmurhash` expect seed to be number) - I would to spend bit more time on digging into it - there was likely cast going on - so I would want to use whatever string was cast into to not cause change in hash generation

zachtylr21 replied:
I thought this was interesting too. It looks like the only place the second argument `seed` is used in the `murmurhash` function is on line 6:

```let h: number = seed ^ str.length```

I tried treating `'abc'` as a hex value, and that doesn't produce the same result. I also tried getting the binary representation of `'abc'` and converting it to a decimal, and that didn't work either. 

I'm not sure what type of casting is going on, but I tested this line several times and it seems like `^ 'abc'` is the identity operator. Which is strange, because why include the XOR if you could just write `let h: number = str.length`?

zachtylr21 replied:
I just noticed that the `murmurhash` function was taken from [here](https://gist.github.com/raycmorgan/588423v), as indicated in the comment at the top of `murmur.ts`. That would explain why that line `let h: number = seed ^ str.length` is there despite there being a simpler alternative. I think this hash function is a general-use implementation of the MurmurHash2 algorithm, and it just so happens that in our case omitting the seed doesn't change anything.

pvdz replied:
Fwiw, `^` is a bitwise op that will force coerce its arguments to a 32bit number, or `0` if it's not a number (explicitly, it will not cause `NaN`). Hence you can throw anything at it, or any other bitwise op, and always be guaranteed a 32bit number back. The coercion does not try different bases afaik.

pvdz replied:
I double checked this, too, and I think the change of `'abc' -> 0` is fine or at least does not change the status quo.

pieh commented on packages/babel-plugin-remove-graphql-queries/src/index.ts:
```diff
@@ -169,34 +236,38 @@ function getGraphQLTag(path) {
   return {
     visitor: {
-      Program(path, state) {
+      Program(path: NodePath<Program>, state: any): void {
```
can we type `state` to avoid `any`? AFAIK we don't pass any options so generic one should work (if there is one)

zachtylr21 replied:
It's hard to tell what `state` is supposed to be because there aren't many resources that I've found on writing babel plugins. Based on [this](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/babel__traverse/index.d.ts) type file, `state` is either the generic type `S` or `any`, and after some digging it looks like `S` is set to `{}` by default. Are you thinking we should just create our own type for `state`?

zachtylr21 replied:
Update on this. I created a custom type for `state` called `IState` and I get an error:

```
Type '(this: {}, path: NodePath<Program>, state: IState) => void' is not assignable to type 'VisitNodeFunction<{}, Program> | VisitNodeObject<{}, Program> | undefined'.
  Type '(this: {}, path: NodePath<Program>, state: IState) => void' is not assignable to type 'VisitNodeFunction<{}, Program>'.
    Types of parameters 'state' and 'state' are incompatible.
      Property 'file' is missing in type '{}' but required in type 'IState'.
```
I think it's because the type of state is expected to be `{}`. However, when I explicitly declare the type like this:
```
Program(path: NodePath<Program>, state: {}): void {
  ...
}
```
I get an error saying `property 'file' does not exist on type '{}'`. I'm unsure how to resolve this issue at the moment. If you know a solution, @pieh, feel free to just add it in.

pvdz commented on packages/babel-plugin-remove-graphql-queries/src/index.ts:
```diff
@@ -169,34 +249,38 @@ function getGraphQLTag(path) {
   return (
     (path.node.callee.type === `MemberExpression` &&
       path.node.callee.property.name === `useStaticQuery` &&
-      path.get(`callee`).get(`object`).referencesImport(`gatsby`)) ||
-    (path.node.callee.name === `useStaticQuery` &&
-      path.get(`callee`).referencesImport(`gatsby`))
+      (path.get(`callee`).get(`object`) as NodePath).referencesImport(
+        `gatsby`,
+        ``
+      )) ||
+    ((path.node.callee as Identifier).name === `useStaticQuery` &&
+      path.get(`callee`).referencesImport(`gatsby`, ``))
```
As an aside, I think the `callee` could be `foo[useStaticQuery]` and this check would pass because it should check `callee.computed` too, but that's a different matter :p

pvdz commented on packages/babel-plugin-remove-graphql-queries/src/index.ts:
```diff
@@ -169,34 +249,38 @@ function getGraphQLTag(path) {
+          JSXIdentifier(path2: NodePath<JSXIdentifier>): void {
             if (
-              [`production`, `test`].includes(process.env.NODE_ENV) &&
+              [`production`, `test`].includes(process.env.NODE_ENV || ``) &&
```
Nit: this could be `??` rather than `||`

pvdz replied:
(same for a few lines lower)

zachtylr21 replied:
This line was changed in another pull request and is now outdated.

pvdz commented:
btw, is there no better way to handle the `.get` method and the `NodePath` casts all over the place? That's a little annoying.

zachtylr21 commented on packages/babel-plugin-remove-graphql-queries/src/index.ts:
```diff
@@ -1,10 +1,72 @@
 class StringInterpolationNotAllowedError extends Error {
-  constructor(interpolationStart, interpolationEnd) {
+  interpolationStart: ISourcePosition
+  interpolationEnd: ISourcePosition
```
True, I just changed this so these variables have the type `ISourcePosition | undefined`

zachtylr21 commented:
@pvdz I find that a little annoying as well, but unfortunately I don't think there is anything we can do about it. The `.get` method returns multiple types so casting the result is necessary. I suppose we could break the result into a separate line, so instead of 

```typescript
return (tag.get(`property`) as NodePath<Identifier>).node.name === `graphql`
```

we have
 
```typescript
const property: NodePath<Identifier> = tag.get(`property`)
return (property).node.name === `graphql`
```

but I don't know if that's much better.

pvdz commented:
Ok well the reported errors seem legit errors. So please run `yarn typecheck` locally and for anything that pops up :)

pvdz commented:
@zachtylr21 friendly ping. Do you want to keep working on this? I think it was close to being merged?

zachtylr21 commented:
@pvdz Yes I'd like to finish this and get it merged. Apologies, I didn't see your comment from 6 days ago. Is running `yarn typecheck` the last thing to do?

pvdz commented:
I usually do `yarn typecheck` and `yarn lint` (and watch for _errors_) because for some reason sometimes errors sometimes only get caught by one or the other. That may have been fixed by now but that's what I still do :)

zachtylr21 commented:
@pvdz I'm getting an error that says `Cannot find module 'gatsby'`. Do I need to import gatsby in a different way?

zachtylr21 commented:
@pvdz There are also a bunch of linting errors in the `gatsby-admin` package, which is outside the scope of this PR.

pvdz commented:
The gatsby-admin stuff is either brand new or old problems that require a rebase/merge of master to resolve. I'll try that.

pvdz commented:
Some `yarn typecheck` stuff left it seems
![image](https://user-images.githubusercontent.com/209817/89940079-40f03a80-dc19-11ea-8585-c0f879e50aec.png)

zachtylr21 commented:
I'm a little confused at how to alter the `graphql` type to include those functions. Here is the definition of the `graphql` type, from `gatsby/index.d.ts`.

![Screen Shot 2020-08-11 at 10 26 21 PM](https://user-images.githubusercontent.com/19400680/89971757-b8c26300-dc21-11ea-8a04-ce73dd8707b0.jpg)

It seems I need to alter the file `gatsby/index.d.ts`. Is that outside the scope of this PR?

Also, I'm not getting the errors on lines 249, 300, and 360. `.unshiftContainer` is defined in `@types/babel__traverse/index.d.ts` and `path.node.callee.property` is type `any` according to `@babel/types/lib/index.d.ts`, so `path.node.callee.property.name` should be valid.

vladar commented on packages/babel-plugin-remove-graphql-queries/src/index.ts:
```diff
@@ -1,10 +1,60 @@
-const graphql = require(`gatsby/graphql`)
-const { murmurhash } = require(`./murmur`)
-const nodePath = require(`path`)
+import { graphql } from "gatsby"
```
I don't think this is a correct change. This imports "graphql" tag from gatsby, while the original line `require("gatsby/graphql")` actually imported graphql tools, including `parse`, etc.

```suggestion
import graphql from "gatsby/graphql"
```

vladar replied:
To clarify, `import graphql from "gatsby/graphql"` imports everything from this file: https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/graphql.js

But `import { graphql } from "gatsby"` imports this function: https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/cache-dir/gatsby-browser-entry.js#L92

zachtylr21 replied:
That's a great catch

zachtylr21 commented:
You guys will have to excuse my ignorance on a couple of these things, I'm pretty new to contributing to Gatsby.

After changing `import { graphql } from "gatsby"` to `import graphql from "gatsby/graphql"` like @vladar suggested, I'm getting a new error.

![Screen Shot 2020-08-13 at 8 14 18 PM](https://user-images.githubusercontent.com/19400680/90201901-cbfc3c80-dda1-11ea-93b2-f623e8b131ee.jpg)

I thought this meant I have to do `npm install` in the `packages/gatsby` directory, but doing that gave me more errors:

![Screen Shot 2020-08-13 at 8 19 44 PM](https://user-images.githubusercontent.com/19400680/90202113-5c3a8180-dda2-11ea-97fd-bbead2029f37.jpg)

I'm super confused as to why none of these packages can be found. Is there something I'm missing here? I really want to get this PR moving along but I keep running into frustrating issues like this.

vladar commented:
I pulled your branch, ran `yarn bootstrap` in the monorepo root and I don't see this problem. I see several other errors though:

```
packages/babel-plugin-remove-graphql-queries/src/index.ts:249:33 - error TS2339: Property 'name' does not exist on type 'ArrayExpression | ArrowFunctionExpression | AssignmentExpression | AwaitExpression | ... 40 more ... | PrivateName'.
  Property 'name' does not exist on type 'ArrayExpression'.

249       path.node.callee.property.name === `useStaticQuery` &&
                                    ~~~~

packages/babel-plugin-remove-graphql-queries/src/index.ts:300:20 - error TS2339: Property 'unshiftContainer' does not exist on type 'NodePath<Program>'.

300               path.unshiftContainer(`body`, importDeclaration)
                       ~~~~~~~~~~~~~~~~

packages/babel-plugin-remove-graphql-queries/src/index.ts:360:20 - error TS2339: Property 'unshiftContainer' does not exist on type 'NodePath<Program>'.

360               path.unshiftContainer(`body`, importDeclaration)
                       ~~~~~~~~~~~~~~~~


Found 3 errors.
```

CI fails with [exact same error](https://app.circleci.com/pipelines/github/gatsbyjs/gatsby/48115/workflows/74c3ddcf-789b-49ac-9621-cb764b1a8fa2/jobs/491305).

So my best guess is that your dependency setup is messed up. Are you using npm? This may be the issue. If so I suggest you remove `node_modules`, run `yarn` in the monorepo root and try again.

zachtylr21 commented:
Ah, I was using `npm`. 🤦  I switched to yarn and now I'm seeing the same errors. I fixed them, and just pushed. I should note that the method `unshiftContainer` doesn't exist in any of Babel's type declaration files, so I had to change that method call to something that I believe is equivalent based on [this](https://stackoverflow.com/questions/37148869/using-babel-how-can-i-append-some-code-to-the-top-of-every-file) StackOverflow post.

pvdz commented:
It seems something snapshotty is failing over this thing https://github.com/gatsbyjs/gatsby/commit/fc976699e856b56e029bbd772282e996d46598e8#diff-0cee23a635d1c08646f35342b6dee83f

pvdz reviewed (approved):
lgtm and tests are passing. Thank you :purple_heart:
